protocols/network.md

```diff
@@ -52,16 +52,16 @@ With one or more IP addresses for the host the client wants to contact, it
 sends a "connect request". The connection it wants to establish is called a
 TCP ([Transmission Control
 Protocol](https://en.wikipedia.org/wiki/Transmission_Control_Protocol)) or
-[QUIC](https://en.wikipedia.org/wiki/QUIC) connection and they work similar to
+[QUIC](https://en.wikipedia.org/wiki/QUIC) connection, which is like
 connecting an invisible string between two computers. Once established, the
 string can be used to send a stream of data in both directions.
 
 If the client has received more than one address for the host, it will
-traverse that list of addresses when connecting and in case one of fails it
-will try to connect to the next one until either one connect works or they all
-fail.
+traverse that list of addresses when connecting, and if one address fails it
+will try to connect to the next one, repeating until either one address works 
+or they've all failed.
```

Comment:
surely this should be "or they all failed" ? The *have* in there doesn't seem necessary.

Reply:
"they all failed" is in the past tense, whereas "one address works" and "they have all failed" are in present perfect tense, which is consistent. Does that make sense? [Here's some explanation](https://pasttenses.com/fail-past-tense).

Reply:
I'm not sure I agree, but I won't insist.